### ISSUE TITLE
ethernet devices crashes under cygwin64b, runs fine under cygwin32b

### ISSUE THREAD

**Issue by AK6DN** [simh/simh#827]
Title says it all. I have an existing 2.11BSD image for the PDP11 that runs perfectly fine under SIMH compiled under cygwin 32b environment. I just installed cygwin 64b environment and recompiled (with no errors) and ran a number of PDP11 test programs I have (user written diagnostics, mainly). They all worked fine, same results on 64b as on 32b cygwin. Running on a 64b Win7 platform.

However, when I ran my 2.11BSD image in cygwin 64b PDP11 it crashes during startup of the networking.  I enabled debug to stderr for the XU interface and captured the console log.

Here is the generated 'pdp11.exe.stackdump' file:
[pdp11.exe.stackdump.txt](https://github.com/simh/simh/files/4395366/pdp11.exe.stackdump.txt)

Here is the full debug console log:
[cpu1144_211bsd_64b_boot_crash.txt](https://github.com/simh/simh/files/4395359/cpu1144_211bsd_64b_boot_crash.txt)

And here is the .ini file I use to startup PDP11 SIMH:
[cpu1144.ini.txt](https://github.com/simh/simh/files/4395369/cpu1144.ini.txt)

**Comment by AK6DN** [comment]
Here are my SIMH PDP11 32b and 64b build logs...
[simh_32b_build_log.txt](https://github.com/simh/simh/files/4395522/simh_32b_build_log.txt)
[simh_64b_build_log.txt](https://github.com/simh/simh/files/4395523/simh_64b_build_log.txt)

**Comment by markpizz** [comment]
This is one of those Guys says: "Doctor it hurts when I do this!"...  Doctor says "Don't do that!"

As I've said in the past, if simh happens to work under any form of cygwin, that's great.  If it doesn't there are at least a half dozen build choices on Windows that do work, and there are the pre-compiled binaries.

**Comment by AK6DN** [comment]
Yup I have heard that before. Wanted to file a report anyway. Cygwin Inc is starting to deprecate the 32b environment and pushing everyone to 64b, so thought I'd try it out. Looks like the issue is specific to the WinPCAP/ethernet subsystem (which itself is old and crusty and basically deprecated).

Tried the Windows binaries, but they want to run in their own CMD console window. Can't even launch them thru the cygwin terminal. They want their own CMD window. That really sucks.

Maybe I'll try loading MINGW as cygwin supports that. I'll see how it works.

BTW, is the future direction to move from WinPCAP to Npcap?

**Comment by markpizz** [comment]
> BTW, is the future direction to move from WinPCAP to Npcap?

That "was" the plan, and for quite some time the code actually had preferred Npcap to WinPcap, when both were installed which was in possible earlier Npcap versions.  However about 6 weeks back it was discovered that Npcap wasn't a 100% functional replacement for WinPcap.  In particular, with Npcap  a simulator couldn't talk to the host system's network stack, while that "just worked" without any sort of bridge monkey business that was necessary on *nix hosts.

There is a a pull request pending in the Npcap project that might fix the problem.  That's been sitting there for 11 days.  Building Npcap from scratch isn't a trivial process, so I'm waiting for someone on the Npcap team to formally adopt that pull request and release it.  If that actually fixes the problem, Npcap will be the future Windows pcap package.

**Comment by AK6DN** [comment]
Yes I saw that issue with Npcap also. 

In looking at the 64b build capture I don't see any 64b library reference .a files. There is an 'x64' subdir under .../WpdPack/Lib, but there are no .a files, only .lib. And cygwin 64b build does not seem to detect or use the x64 directory. I am not familiar enough with WinPCAP to know if this is an issue.

But I found an old simple tool I had compiled back about 5 yrs ago under cygwin 32b to find and itemize all the detected interfaces no longer compiles without error under cygwin 32b or 64b. Blah.

Maybe I'll just move to linux. I'm not looking forward to Win10.

**Comment by markpizz** [comment]
For a VERY LONG time (11 or so years), pcap Ethernet support on all platforms has loaded the host platform pcap library at runtime.  At compile time only the pcap.h file is used to define the API interfaces.  No link libraries are needed or wanted.  USE_SHARED selects this behavior.

**Comment by AK6DN** [comment]
Ok that is an interesting piece of knowledge I was not aware of. I just assumed libwpcap.a and libpacket.a were required to link against. There are 32b versions but none in the 64b directory.

**Comment by markpizz** [comment]
They should merely be remnants of earliest functionality.  Try renaming them and see that things build and run successfully.  Let me know.

- Mark

**Comment by AK6DN** [comment]
Well, I finally got it no at least not crash on startup.
Under windows-build/winpcap/WpdPack/Include I left all the standard .h files. So as distributed.
Under windows-build/winpcap/WpdPack/Lib I left (only) a zero-sized libwpcap.a file.

Both the 32b and 64b versions compile OK with this. 32b works fine, as before.
64b sort of works. Does not crash, but I can only see/ping 127.0.0.1 and 192.168.1.11 which is the static assigned IP address to the de0 interface.

Well I spoke too soon, I guess. It now crashed after about 10min or so:

64b[15] ping 192.168.1.11
PING 192.168.1.11 (192.168.1.11): 56 data bytes
64 bytes from 192.168.1.11: icmp_seq=0 ttl=255 time=16.667 ms
64 bytes from 192.168.1.11: icmp_seq=1 ttl=255 time=16.667 ms
64 bytes from 192.168.1.11: icmp_seq=2 ttl=255 time=16.667 ms
^C
--- 192.168.1.11 ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 16.667/16.666/16.667 ms
64b[16] /sbin/ifconfig de0
de0: flags=63<UP,BROADCAST,NOTRAILERS,RUNNING>
        inet 192.168.1.11 netmask ffffff00 broadcast 192.168.1.255
64b[17] Segmentation fault (core dumped)
simh[510]

So still some more work to do, I guess.
It is weird to me tho that:
windows-build/winpcap/WpdPack/Lib/libwpcap.a
seems to need to exist to allow make to proceed, but it is not used in the build command.

**Comment by AK6DN** [comment]
Well, given the new release of NPCAP 0.9990 thought I would try that.
Uninstalled WinPCAP 4.1.3, installed NPCAP 0.9990. That went OK.
WireShark actually works fine.
Rebuilt PDP11 under Cygwin32b. Launched and loaded 2.11BSD image. Network works fine.
Rebuilt PDP11 under Cygwin64b. Launched and loaded 2.11BSD image. Crashes almost immediately.
Does not even get as far as with using WinPCAP 4.1.3. Sigh.
I reused the existing WinPCAP Include files when building. Maybe that is an issue?
The build ran with no errors. I'll try gloming in the the NPCAP SDK Include files, but I am not hopeful at this point it will make a difference.
So it looks like Cygwin64b is a NOGO for xPCAP use at this time.

**Comment by AK6DN** [comment]
Ok, substituted Include files from 'npcap-sdk-1.05.zip' in place of the standard windows-build winpcap include files. Both 32b and 64b pdp11.exe images built with no errors. 32b image works correctly using 2.11BSD disk image. 64b image crashes the same as before. Basically no difference.
So my conclusion is still the same:  looks like Cygwin64b is a NOGO for xPCAP use at this time.

**Comment by AK6DN** [comment]
Tried same cygwin pdp11.exe 32b and 64b images under Win10Pro64b rel 1909 and similar behavior.
32b runs 2.11BSD image fine, networking works as expected.
64b fails, but does not crash. Just does nothing. ping's from console timeout, no response.
Not crashing is an improvement, but still not working either.
This was using NPCAP 0.9990 installed on Win10.
